domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt

```diff
@@ -47,23 +59,38 @@ class ExplorationProgressController @Inject constructor(
   //  to avoid cases in tests where the exploration load operation needs to be fully finished before
   //  performing a post-load operation. The current state of the controller is leaking this
   //  implementation detail to tests.
+  // TODO(): Update the saving mechanism for checkpoints once internal locking of this controller is
```

Comment:
Please add an issue for this TODO or resolve it.

Reply:
@BenHenning In the earlier approach, we had to acquire `exploration progress lock` twice, first to create a checkpoint and then again to update the `checkpointState` once the save operation has completed. This created a race condition that if the user exits the exploration after the save operation has started but has not finished, we might end up with the wrong `checkpointState`.

But with the current approach, we kick off the saving process, wait for it to complete and then update the checkpoint state without releasing the lock, so I think this eliminates the race condition. WDYT?

Reply:
I don't think it does--we're still waiting for the save checkpoint Deferred to finish and we don't block the DataProvider on that. Things could change between submit answer and the saving completed which could introduce more data races. Ideally, we should ensure eventual consistency (i.e. regardless of the order of operations, the final outcome is still the correct one).

Reply:
@BenHenning  I have one question out of curiosity. We are waiting for the exploration for the save to complete before advancing the exploration to viewing state doesnt that mean we are blocking the dataProvider.

Reply:
added a TODO and filled a new issue for this

Reply:
@MaskedCarrot that would be the case if not for the gap. The current flow is essentially:

- Main thread (from UI class): perform some state change (such as submitting an answer)
- Main thread (from progress controller): initiate save; acquire lock
- Main thread (from checkpoint controller): call storeDataWithCustomChannelAsync
- Main thread (from progress controller): call invokeOnCompletion
- Main thread (from progress controller): return back to UI class with DataProvider corresponding to operation; release lock (not exactly; the critical section is smaller than indicated here but this better illustrates the point)
- Background thread A (from checkpoint controller): initiate save operations, then finish
- Background thread A (from progress controller via invokeOnCompletion callback): acquire lock, retrieve results & notify UI layer that the state data provider has changed, release lock

Notice that two different threads are acquiring the lock at different times. The time it takes from when the UI thread releases the lock to invokeOnCompletion being called is potential for a race condition to occur.

Reply:
Understood. Thanks for the explanation.